File: one_fm/api/doc_methods/help_category.py
import frappe
from frappe import _
from slugify import slugify


def validate(doc, method):
    """
        Validate data
    """
    if doc.category_name == doc.category:
        frappe.throw(_("Category name should not be same as sub-category."))
    # check for sub category
    if (doc.category and frappe.get_doc("Help Category", doc.category).is_subcategory):
        frappe.throw(_("Subcategory cannot be selected as category."))

def before_insert(doc, method):
    """
        Rename route before saving in help category
    """
    # set route
    if (doc.is_subcategory and doc.category):
        doc.route = f"/knowledge-base/{slugify(doc.category, allow_unicode=True)}/{slugify(doc.category_name, allow_unicode=True)}"
    else:
        doc.route = f"/knowledge-base/{slugify(doc.category_name, allow_unicode=True)}"

File: one_fm/public/js/doctype_js/help_category.js
frappe.ui.form.on('Help Category', {
    refresh:function(frm){
      // set query
      set_query(frm)
    }
});

// SET QUERY
let set_query = frm => {
  // filters category for notself and not sub category
  frm.set_query('category', () => {
      return {
          filters: {
              name: ['!=', frm.doc.name],
              is_subcategory: 0
          }
      }
  })
}
